core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -140,12 +141,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     }
 
     override def receive: PartialFunction[Any, Unit] = {
-      case StatusUpdate(executorId, taskId, state, data, resources) =>
+      case StatusUpdate(executorId, taskId, state, data, taskCpus, resources) =>
         scheduler.statusUpdate(taskId, state, data.value)
         if (TaskState.isFinished(state)) {
           executorDataMap.get(executorId) match {
             case Some(executorInfo) =>
-              executorInfo.freeCores += scheduler.CPUS_PER_TASK
+              executorInfo.freeCores += taskCpus
```

Comment:
I put it there so that I don't have to worry about race conditions between when we get status updates from executors and when the scheduler removes the tasksets. for instance currently we could remove it from the taskidToTaskSetManager if the executor failed.  In the case the taskid to task set manager is missing we wouldn't know how many cores to properly add back. So to keep away from that race this seems cleaner and also possibly gives us a way to expose to the user in the future if we wanted.

Reply:
Oh..oops..I copied example from your code without update and leave you to another direction. Sorry about that.

Actually, I mean, could we do this?

`sc.resourceProfileManager.taskCpusForProfileId(executorInfo.resourceProfileId)`

Get resourceProfileId from executorInfo instead of taskSet.

Reply:
ah, yes we can get it from there. we aren't using it for anything else at this point.